dimod/include/dimod/quadraticiterator.h

```diff
@@ -0,0 +1,218 @@
+// Copyright 2020 D-Wave Systems Inc.
+//
+//    Licensed under the Apache License, Version 2.0 (the "License");
+//    you may not use this file except in compliance with the License.
+//    You may obtain a copy of the License at
+//
+//        http://www.apache.org/licenses/LICENSE-2.0
+//
+//    Unless required by applicable law or agreed to in writing, software
+//    distributed under the License is distributed on an "AS IS" BASIS,
+//    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//    See the License for the specific language governing permissions and
+//    limitations under the License.
+
+#ifndef DIMOD_QUADRATICITERATOR_H_
+#define DIMOD_QUADRATICITERATOR_H_
+
+#include <iterator>
+#include <type_traits>
+#include <utility>
+
+namespace dimod {
+namespace utils {
+    template <class V, class B>
+    struct QuadraticTerm {
+        V const u;
+        V const v;
+        B const bias;
+
+        QuadraticTerm const* operator->() const { return this; }
+    };
+
+    template <class InIT, class OutIT>
+    class QuadraticIterator {
+      public:
+        using in_iterator = InIT;
+        using out_iterator = OutIT;
+
+        using bias_type = typename std::remove_cv<typename out_iterator::value_type::second_type>::type;
+        using variable_type = typename std::remove_cv<typename out_iterator::value_type::first_type>::type;
+
+        using difference_type = typename out_iterator::difference_type;
+        using value_type = QuadraticTerm<variable_type, bias_type>;
+        using pointer = value_type const*;
+        using reference = value_type;
+        using iterator_category = std::input_iterator_tag;
+
+        QuadraticIterator() = default;
+        QuadraticIterator(in_iterator in_it, in_iterator in_end, variable_type u = 0)
+            : u_(u), in_it_(in_it), in_end_(in_end) {
+            if (in_it_ != in_end) {
+                out_it_ = out_begin();
+                operator++();
+            }
+        }
+
+        reference operator*() const { return {u_, out_it_->first, out_it_->second}; }
+        reference operator->() const { return operator*(); }
+
+        QuadraticIterator& operator++() {
+            ++out_it_;
+            while (out_it_ == out_end() || u_ < operator*().v) {
```

Comment:
When there are two conditions, its better to put the more probable one first, so that the compiler skips evaluating two conditions. in this case u_ < operator*().v is more probable of evaluating to true than out_it_ == out_end() , but then again how these two checks will be ordered in the end by the compiler is not too predictable, but its better to be on the safe side for frequently used functions like this.

Reply:
The two conditions are checked in this order to prevent dereferencing the iterator if it is past the end.